### DISCUSSION TITLE
softcut <-> scsynth routing

### DISCUSSION THREAD

**Discussion by catfact** [monome/norns#1683] [General]
it's coming up again.

https://llllllll.co/t/fx-mod/62726/10?u=graymazes

there is, and always shall be i believe, a persistent desire to route softcut output into scsynth, and vice versa, and maybe sometimes both at once (despite the possiblity of bad feedback)

we've discussed this in various places but have acheived no consensus. so here i am trying again.

the current practice is for scripts to sometimes reroute the JACK client connections. this is kind of ok, but kind of bad because we have no visibility on this state through the norns system, and no real mechanism to e.g. reset  it to a known condition.

i believe this should be a system API. i would not suggest that the system API touch the JACK client routing - we already have a single "leader" client  (the `MixerClient`) through which all other client I/O is routed. so we can simply add connections here:

https://github.com/monome/norns/blob/main/crone/src/MixerClient.cpp#L32-L39

and have this connected to a stateful API that, say, knows three states (`cut->ext`, `ext->cut`, `cut<->ext`) and always defaults to `ext->cut`.

**Comment by catfact** [comment]
i sort of feel like we don't really need a levels control on the mix page for this. just three states. and perhaps not even a UI access point, only a scripting API. (does anybody ever set `ext->cut` to any level except silence or unity?)

**Reply by catfact** [reply]
making it a toggle would just be a minor DSP performance optimization

**Reply by RajaTheResidentAlien** [reply]
(nvm what i wrote above, i understand now) just making it a toggle sounds good 👍

**Comment by catfact** [comment]
piggybacking on this, there's been interest expressed in different pre/post options for (e.g.) hearing the engine signal fully wet in the reverb. (not currently possible. not something i would do but it's a simple change.)

what other routing desires have come up that aren't currently supported?

**Comment by RajaTheResidentAlien** [comment]
(that's strange: there's already reverb-capability within the engine
itself... i hope softcut can still get some effects-love from the
engine-layer, this seems most imperative to me :D)

within just the API, i would love if "audio.level_eng_cut" worked similar
to "softcut.level_input_cut" whereby the first argument is the
supercollider-engine input bus-number('1' or '2'), the second argument is
the softcut voice number to be routed to specifically('1' through '6'), and
the third argument is the amp-volume/level('0' - '11' ;D ...just kidding,
whatever you think is reasonable, but i notice i'm able to amp
'softcut.level_input_cut' at values higher than 1, found it useful for
quiet inputs).
basically, i want to be able to route different busses of the engine layer
into different voices of softcut, more specifically.

thank you for considering all this :)

**Comment by sonocircuit** [comment]
thanks for picking this up again!

personally I'd be very happy to have pre/post reverb toggles for engine, sofutcut, monitor and tape and a toggle to switch between supercollider > softcut / softcut > supercollider routings.
